core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -2007,8 +2007,13 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
       "offsets.commit.required.acks must be greater or equal -1 and less or equal to offsets.topic.replication.factor")
     require(BrokerCompressionCodec.isValid(compressionType), "compression.type : " + compressionType + " is not valid." +
       " Valid options are " + BrokerCompressionCodec.brokerCompressionOptions.mkString(","))
-    require(!processRoles.contains(ControllerRole) || controllerListeners.nonEmpty,
-      s"${KafkaConfig.ControllerListenerNamesProp} cannot be empty if the server has the controller role")
+    if (usesSelfManagedQuorum) {
```

Comment:
A couple additional validations we can do:

1. Ensure control plane listener is empty. We have this check in BrokerServer already, but we should check it here as well.
2. Verify port consistency between the controller listener and the quorum voter string. As you noted offline, the port gets duplicated, so we should be sure they match. 
3. We currently assume that the first listener in `controller.listener.names` is the one that matches the voter string. We should either require this to be the case, or change the logic in `RaftManager` to choose the right listener.

Reply:
> 2. Verify port consistency between the controller listener and the quorum voter string. As you noted offline, the port gets duplicated, so we should be sure they match.

One complexity is when the controllers let the OS assign a port (`0`) to the listener. This feature is heavily used in the tests to allow multiple controllers to run on the same host.

Reply:
I think that's fine. When the test cases override `controller.quorum.voters`, they also override the controller listener. For example, see here: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala#L220. So it seems reasonable to validate port consistency in any case.

Reply:
Okay. How about this algorithm/check:

If `process.roles=broker` then Kafka requires `controller.listener.names` is set to one name and doesn't allow the `listeners` to contain `controller.listener.names`. This is consistent since controller ports are only specified in `controller.quorum.voters`

When `process.roles=controller,*` then Kafka requires `controller.listener.names` is set and `listeners` contains `controller.listener.names`. Additionally check that the port in `controller.quorum.voters` for `node.id` matches the port in the `listeners` for one of `controller.listener.names`. Fix `RaftManager` to pick the right controller listener name based on the port of it's id in the voters.

Reply:
I'm a little unsure about requiring the broker have only one controller listener. Technically it's an incompatible change. Another way we could handle it is to log a warning that other specified ports will not be used. 

The second part sounds good to me.

Reply:
I'm in agreement with not allowing `listeners` to contain the `controller.listener.names` values(s).  I'll rework tghe PR to do that.

FYI, just as a sanity check, I confirmed that I was able to get a separate controller and broker working using our QuickStart files with the following changes:

**controller.properties**
```
listeners=SASL_PLAINTEXT://:9093
listener.name.sasl_plaintext.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule Required unsecuredLoginStringClaim_sub="broker";
sasl.enabled.mechanisms=OAUTHBEARER
sasl.mechanism.controller.protocol=OAUTHBEARER
controller.listener.names=SASL_PLAINTEXT
```

**broker.properties**
```
listener.security.protocol.map=CONTROLLER:SASL_PLAINTEXT,PLAINTEXT:PLAINTEXT,SSL:SSL,SASL_PLAINTEXT:SASL_PLAINTEXT,SASL_SSL:SASL_SSL
sasl.mechanism.controller.protocol=OAUTHBEARER
listener.name.controller.oauthbearer.sasl.jaas.config=org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule Required unsecuredLoginStringClaim_sub="broker";
```

I was able to do this both with requiring the additional listener on the broker (the initial version of this PR) and not (the way the code works without this PR).

So while the "listener" terminology is unfortunate in that the broker is not "listening" on it, I have confirmed that everything works as expected.

I'll work on this more tomorrow.

Reply:
@rondagostino How about the 3 additional validations I mentioned in the initial comment above?

Reply:
Pushed a commit that does the following:

- Ensure control plane listener is empty
- If process.roles=broker then don't allow the listeners to contain controller.listener.names
- When process.roles=controller,* then require controller.listener.names is set and listeners contains controller.listener.names

Still need to verify port consistency between the controller listener and the quorum voter string.  There is some subtlety regarding what the right thing to do is on a broker-only node vs one that has the controller node, which I will have to think about a bit.

Reply:
Ok, I pushed a commit that checks the consistency of the ports between the `controller.quorum.voters` and (`controller.listener.names`/`listeners`) configs.  This can only be done for the controller case since there is nothing in `listeners` for the broker-only case.

Currently both `RaftManager` and `RaftControllerNodeProvider` take `config.controllerListenerNames.head` to determine how to connect.

`RaftControllerNodeProvider` extends `ControllerNodeProvider` which exposes a single listener name/security protocol/SASL mechanism.  So we aren't going to make `RaftControllerNodeProvider` sophisticated about which security protocol/SASL mechanism to use based on voter ports -- it has to always use a single one.  This makes sense given that migrating communication to the ZK-based brokers (any one of which can become the controller) from one security protocol to another involves adding the second security protocol listener everywhere, migrating everybody over to using that second security protocol (e.g. via `inter.broker.security.protocol`), and then dropping the first listener -- so 3 rolls.  We would do something similar in KRaft: add the second controller listener on all controllers, migrate all brokers to use it (by making that listener the first one in the controller listener names list), then drop the original listener on the controllers.

It now becomes appropriate to discuss `RaftManager`.  `RaftManager` is used on a broker-only node, and in that case it is not going to be unable to use the ports in `controller.quorum.voters` to identify a security protocol because there is nothing in `listeners` to find based on a port number.  So it has to use `config.controllerListenerNames.head` in this case.

`RaftManager` could potentially be sophisticated about which security protocol/SASL mechanism to use based on voter ports in the case where we have the controller role -- there are listeners that it can identify based on port numbers.  But I don't think we would want to in this case, which generally corresponds to the above migration scenario.  I think the way that scenario would work would be as follows:

1. Append a second entry to `controller.listener.names` to all nodes with the controller role plus a corresponding entry in `listeners`, and roll all controllers.  This has no effect other than adding the second active controller listener to each controller -- in particular, they won't hear anything yet.
2. Now that all controllers have both listeners, swap the order of `controller.listener.names` and update `controller.quorum.voters` with the new port, and roll all controllers again.  Now the controllers are talking to each other on that new listener/security protocol, and any brokers running in combined mode are as well.
3. Set `controller.listener.names` to be the new listener name and update `controller.quorum.voters` on any broker-only nodes and roll them.  Now the broker-only nodes are also talking to the controllers over the new security protocol.
4. Eliminate the old entry from `controller.listener.names` (it is now the second in the list at this point, and we are deleting it) on any controller node and roll the controllers a third time.  We have completed the migration.

So it takes 4 rolls: 3 rolls of the controller nodes and one roll of the broker-only nodes.

The point here is that we never have a situation where a running instance of a broker or a controller is talking to some controllers with one security protocol and others with another security protocol.  Every node always talks to all controllers over the same security protocol at any particular moment.

Assuming this all makes sense and we agree, I think this has the following implications:

* I added validation of port consistency for controllers by checking that the port existed in at least one controller listener.  I think this validation should be more strict: the port must match the first controller listener listed.
* For broker-only nodes, `controller.listener.names` should only have a single entry.  Given the point above that this is technically an incompatible change, I think the suggestion to log a warning is appropriate.

Thoughts?

Reply:
I've defined the voter port validation to check that this controller's voter node port matches the first controller listener port.  We can't validate all controllers' voter ports because we run multiple controllers on the same host in tests, and in that case every controller must list on a different port despite using the same security protocol.

Reply:
I've removed voter port validation because an offline discussion pointed out that we can't validate the `controller.quorum.voters` ports (which are the ports that clients use and are semantically "advertised" ports even though the controller configuration doesn't list them in advertised.listeners) against the `listener` ports (which are semantically different then the ports that clients use).